### ISSUE TITLE
Variable defined with walrus operator in comprehension incorrectly flagged as unresolved reference

### ISSUE THREAD

**Issue by Andrej730** [astral-sh/ty#3077]
### Summary

Hi! Stumbled upon `ty` incorrectly flagging a variable defined with a walrus operator usage inside a list comprehension as an unresolved reference, even though the variable is defined outside of the comprehension.

```python
lengths = list(range(5))
cumsum = 0
cs = [cumsum := cumsum + x for x in lengths]
print(cs)     # [0, 1, 3, 6, 10]
print(cumsum) # 10
```

Playground - https://play.ty.dev/d93a77d9-98f7-4326-874c-ad12e171b8bc

```
error[unresolved-reference]: Name `cumsum` used when not defined
 --> test.py:3:17
  |
1 | lengths = list(range(5))
2 | cumsum = 0
3 | cs = [cumsum := cumsum + x for x in lengths]
  |                 ^^^^^^
4 | print(cs)
```

### Version

ty 0.0.23 (be6e94a27 2026-03-13)

**Comment by AlexWaygood** [comment]
Thanks for the report! Definitely a bug. This might be the same thing as https://github.com/astral-sh/ty/issues/162, but I'm not totally sure.

**Comment by carljm** [comment]
It is just an aspect of #162. Without that special handling for walrus assignments, we treat `cumsum` as a name bound in the comprehension scope. As a local of a function-like scope, uses cannot walk up to enclosing scopes. If the name is accessed before it's written in that scope, that's just an `UnboundLocalError` at runtime, and that's what ty models (correctly -- if it weren't for the special case for walrus assignments in comprehensions.)

**Comment by Andrej730** [comment]
Okay, I'll use `cs = list(itertools.accumulate(lengths))` as a workaround in my case.
